### PR TITLE
Fix foreign key and module inference for multi word structs

### PR DESCRIPTION
**Pull request by mcasper**
Change foreign_key and association inference to work for structs/tables with underscores, and set up
User -> SpecialPost -> SpecialComment in the test schema to test it.

### CODE REVIEW

File: diesel_codegen/src/associations/belongs_to.rs [hunk 1]
```diff
@@ -95,6 +95,18 @@ impl BelongsToAssociationBuilder {
...
 }
 
+fn capitalize_from_association_name(name: &str) -> String {
+    let mut result = String::with_capacity(name.len());
+    let words: Vec<&str> = name.split("_").collect();
+
+    for word in words {
+        result.push_str(&word[..1].to_uppercase());
+        result.push_str(&word[1..]);
+    }
```
Comment (sgrif) [inline diesel_codegen/src/associations/belongs_to.rs]: I almost want to write this as:

```rust
name.split("_").map(|word| vec![word[..1].to_uppercase(), word[1..].into()]).collect()
```

But I'm not sure if it's better or not.

File: diesel_codegen/src/associations/belongs_to.rs [hunk 2]
```diff
@@ -95,6 +95,18 @@ impl BelongsToAssociationBuilder {
     }
 }
 
+fn capitalize_from_association_name(name: &str) -> String {
+    let mut result = String::with_capacity(name.len());
+    let words: Vec<&str> = name.split("_").collect();
```
Comment (sgrif) [inline diesel_codegen/src/associations/belongs_to.rs]: `collect` is not needed here.

File: diesel_tests/tests/schema.rs [hunk 3]
```diff
@@ -58,11 +58,55 @@ pub struct Comment {
...
+        NewSpecialComment {
+            special_post_id: special_post_id
+        }
+    }
+}
+
+#[insertable_into(special_comments)]
+pub struct NewSpecialComment {
+    special_post_id: i32
+}
```
Comment (sgrif) [inline diesel_tests/tests/schema.rs]: Are we able to inline these structs into the test functions? This file is starting to become a dumping ground.

Comment (sgrif) [inline diesel_tests/tests/schema.rs]: If the only problem with doing that is errors about privacy, I'm fine with adding some `no_privacy=true` options that are internal only to codegen. I'd also prefer to duplicate these structs if they are only used in 2 tests.

**Comment by mcasper** [comment]
@sgrif If we move the structs out of `schema.rs`, we'll need to make a new mod for the test and add it to `lib.in.rs`, right?

**Comment by sgrif** [comment]
@mcasper Ugh.... Right, stable... I think it's probably fine to move the modules being tested into `lib.in.rs` (I say this quite begrudgingly).

**Comment by sgrif** [comment]
Or at the very least let's just make a module that is `structs_with_codegen_for_testing_foo` be a module.

**Comment by mcasper** [comment]
@sgrif I moved everything into an `annotations` module.

**Comment by sgrif** [comment]
I don't like having a general `annotations` module, as that changes from having `schema` be a generic dumping ground to having `annotations` be a generic dumping ground. Can we at least make these be one module per test if we think that inlining them into the tests themselves is too painful?

**Comment by mcasper** [comment]
Do we like inlining the structs into the tests more because it doesn't allow them to be reused?

**Comment by sgrif** [comment]
It also just makes it clear where they are. I just don't want to end up with https://github.com/rails/rails/blob/master/activerecord/test/models/post.rb

**Comment by mcasper** [comment]
Man, I was really hoping we could reach `very_special_comment` level. I'm totally fine with inlining the structs in those tests, do we also want a more explicit module name?

**Comment by sgrif** [comment]
The only downside to inlining them into the tests is that the modules that the tests are in need to go through syntex on stable (which means no line numbers on compiler errors). Given that this is development only, and most of us (all of us?) who work on the framework use nightly for day-to-day dev, this is probably fine.

**Comment by mcasper** [comment]
Sounds good to me, inlined the structs

**Pull request merged**